rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/HostsFileBasedInfrastructureManager.java

```diff
@@ -269,27 +276,54 @@ public void removeNode(Node node) {
         InetAddress host = null;
         String nodeName = node.getNodeInformation().getName();
         if ((host = registeredNodes.remove(nodeName)) != null) {
-            freeHosts.add(host);
-            logger.debug("Node " + nodeName + " removed. #freeHosts:" + freeHosts.size() + " #registered: " +
-                registeredNodes.size());
-            try {
-                this.killNodeImpl(node, host);
-            } catch (Exception e) {
-                logger.trace("An exception occurred during node removal", e);
+            // remember the node removed
+            removedNodes.putIfAbsent(host, new AtomicInteger(0));
+            removedNodes.get(host).incrementAndGet();
+            if (!registeredNodes.containsValue(host)) {
+                // in case all nodes relative to this host were removed kill the JVM
+                freeHosts.put(host, removedNodes.remove(host).intValue());
```

Comment:
I would have moved this line after `this.killNodeImpl(node, host);` to be sure that a node is deleted and forgotten only if the runtime on the remote node has been really killed, otherwise the information at the Scheduler level may be wrong. 

Same apply for other infrastructures.